src/core/ext/xds/xds_channel_creds.h

```diff
@@ -19,30 +19,45 @@
 
 #include <grpc/support/port_platform.h>
 
+#include <grpc/impl/codegen/grpc_types.h>
+
 #include "src/core/lib/json/json.h"
-#include "src/core/lib/security/credentials/credentials.h"
+
+struct grpc_channel_credentials;
 
 namespace grpc_core {
 
-class XdsChannelCredsImpl {
+class XdsChannelCredsFactory {
  public:
-  virtual ~XdsChannelCredsImpl() {}
+  virtual ~XdsChannelCredsFactory() {}
   virtual absl::string_view creds_type() const = 0;
   virtual bool IsValidConfig(const Json& config) const = 0;
-  virtual RefCountedPtr<grpc_channel_credentials> CreateXdsChannelCreds(
+  virtual grpc_channel_credentials* CreateXdsChannelCreds(
```

Comment:
Why change this from `RefCountedPtr<>` to a raw pointer?  I think `RefCountedPtr<>` makes more sense here.

Reply:
Thanks, I tried but could not figure out how to make it work with forward declaration (due to a circular dependency) here.
Is there a tip?

Reply:
What is the circular dependency you're seeing?  I would expect that you can simply have the BUILD target for "xds_channel_creds" depend on "ref_counted_ptr".

Reply:
That gave me the following error:
./src/core/lib/gprpp/ref_counted_ptr.h:105:34: error: member access into incomplete type 'grpc_channel_credentials'
    if (value_ != nullptr) value_->Unref();
src/core/ext/xds/xds_channel_creds.cc:37:41: note: in instantiation of member function 'grpc_core::RefCountedPtr<grpc_channel_credentials>::~RefCountedPtr' requested here
  if (iter == factories_.cend()) return nullptr;
./src/core/ext/xds/xds_channel_creds.h:26:8: note: forward declaration of 'grpc_channel_credentials'
struct grpc_channel_credentials;

When I put RefCounted<grpc_channel_credentials> instead of grpc_channel_credentials, then it gives:

./src/core/lib/gprpp/ref_counted.h:304:49: error: static_cast from 'grpc_core::RefCounted<grpc_channel_credentials, grpc_core::PolymorphicRefCount, grpc_core::kUnrefDelete> *' to 'grpc_channel_credentials *', which are not related by inheritance, is not allowed
      internal::Delete<Child, UnrefBehaviorArg>(static_cast<Child*>(this));
./src/core/lib/gprpp/ref_counted_ptr.h:105:36: note: in instantiation of member function 'grpc_core::RefCounted<grpc_channel_credentials, grpc_core::PolymorphicRefCount, grpc_core::kUnrefDelete>::Unref' requested here
    if (value_ != nullptr) value_->Unref();
// Same last two lines.

Reply:
BTW by the circular dependency I meant grpc_security_base for grpc_channel_credentials.
config -> xds_channel_creds -> grpc_security_base -> grpc_base -> config

Reply:
(Sorry, I guess I accidentally resolved this comment.)

Reply:
Updated, finally could make it work with templates.

I blocked instantiating the factory with other template types but did not do that with the registry since it's only for CoreConfiguration. But new factories will have to extend XdsChannelCredsFactory<> now and I guess we cannot remove '<>' until we have CTAD in C++17. PTAL.

Reply:
Ah, okay, I see -- the problem is not depending on `RefCountedPtr<>`, but rather the fact that `RefCountedPtr<>` requires the implementation of `grpc_channel_credentials`, which is a larger dependency that we don't want to drag in here.

It's probably better to just use `grpc_channel_credentials*` instead of `RefCountedPtr<grpc_channel_credentials>`.  That's probably simpler than this templating approach.

Reply:
Chat offline, and we decided to keep the current template version.